### ISSUE TITLE
Investigate why Woodwork integration caused issues with Equal and NotEqual primitives

### ISSUE THREAD

**Issue by thehomebrewnerd** [alteryx/featuretools#1584]
### Investigate why Woodwork integration caused issues with Equal and NotEqual primitives

During the Woodwork integration, new issues arose with the `Equal` and `NotEqual` primitives. The error occurred when attempting to compare two categorical series with the categories in the dtype in different orders. This issue was resolve by reordering the categories on one of the two input series:
```
y_vals = y_vals.cat.reorder_categories(x_vals.cat.categories)
```
We should further investigate the root cause of this issue - why does it occur now and did not before? Understanding the performance hit caused by reordering the categories would also be beneficial.

See PR #1568 for reference

**Comment by frances-h** [comment]
After doing some investigation, the main issue is that [pandas does not allow comparisons](https://pandas.pydata.org/docs/user_guide/categorical.html?highlight=categorical#comparisons) between ordered categoricals with different category orderings. 

This leads to a question: how do we want to handle ordinals with different orderings? The current behavior seems to be "treat categoricals as a regular series" and force the comparison by modifying the categories and ordering.

If we want to continue to force the comparison even if the categories do not match and/or do not have the same order, it's faster to just do the comparison on an un-ordered version of the column instead of re-ordering one of the columns. It is also faster to use `set_categories` than `add_categories`. 

If we don't want to force the comparison then there should be a larger discussion on how we want to approach comparisons between categoricals in general.

**Comment by thehomebrewnerd** [comment]
Not exactly sure how I feel about this, but I'll throw out a couple thoughts for consideration.

My general thought is that we should only do item by item comparisons between categorical columns, including ordered and unordered, when the defined categories match. 

The unordered case:
If we have two categorical columns, one with categories of `"a", "b", "c"` and the second with categories of `"a", "b"`, how can we be sure that "a" in the first series represents the same thing as "a" in the second? By defining categories aren't we effectively setting the list of acceptable values? If one series defines three values as acceptable but the other only defines two as acceptable, that to me seems to be a bit of an indication that they represent different things, in which case I'm not sure we should consider items with the same value to be equal.

The ordered case:
Using similar logic for the ordered case, if we have one series with an order of `1, 2` and another with `2, 1`, I'm not sure we can say `1` means the same thing in both series, so perhaps they should not be considered equal?

So basically I think I'm suggesting that if the inputs are categorical with different categories defined we should consider all items between the two to be not equal, so the Equal primitive would return all `False` and the NotEqual primitive would return all `True`.

I'm not convinced that is correct though - certainly interested in other opinions on this.

**Comment by frances-h** [comment]
I'm inclined to agree that we shouldn't be comparing across different categories or orderings. I think we're also going to run into the same issue with LessThan and GreaterThan, but in that case we can't just default to True or False. Both error if you pass in two columns with different orderings. 

Honestly I'm not sure we should even be creating comparison features for columns with differing categories. Maybe we could update primitive matching to include an option that verifies that categories and order match?

**Comment by rwedge** [comment]
Perhaps if a categorical column is not flagged as ordered, we sort the category list so we can avoid Pandas not comparing categorical series with the same set of categories?

**Comment by davesque** [comment]
I'm in favor of what I think @thehomebrewnerd and @frances-h have suggested, which is that we generally shouldn't try to force comparisons of any kind between different/non-existent orderings.  That seems to risk outputting some feature which isn't meaningful.  Seems to me like we should update the matching rules to discard features that would be generated by doing a forced categorization or re-categorization.  By the way, it does seem that this operation could viewed as being relatively expensive.  I was poking around in the code and it seems that `reorder_categories` basically just does an internal call to `set_categories`.  That, in turn, performs re-coding and filtering operations that probably are at least `O(n)` for the series length.

**Comment by thehomebrewnerd** [comment]
@gsheni After further review of this issue, I am unable to reproduce the error that originally prompted the change to the `Equal` and `NotEqual` primitives. These changes are being reverted in PR #1640.

I also propose we delay further work on this issue for now. It seems the Woodwork integration highlighted a potential issue here, but the integration was not the cause of the issue. I believe we should be a bit more deliberate in how we approach updating these primitives. 

We had decided that we should not force comparisons between categorical columns that have different categories or orderings. However, we should also determine if we want to allow comparisons between categorical columns and other columns such as numeric or string columns that do not have categories defined. Similarly, should we allow comparisons between numeric columns and string columns, which will typically result in a feature column of all True or all False values, which is not meaningful.

Disallowing these comparisons will likely require an update to the primitive matching approach, which may prove to be non-trivial. With those items in mind, I think we should delay work on this issue until after the Woodwork integration is complete, and then revisit and prioritize it appropriately later.

**Comment by gsheni** [comment]
@thehomebrewnerd that sounds good. Let's punt on this for now until after Featuretools 1.0.0 is released.

**Comment by thehomebrewnerd** [comment]
This issue is causing failures in the performance test `ManyFeaturesEcommerceTest`. For now, these primitives will be removed from the test, but should be added back in once this issue is closed and resolved.
